proposals/lip-0039.md

```diff
@@ -302,273 +328,365 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
-        let nextQuery = next element of sortedQueries
-        if nextQuery exists and areSiblingQueries(q, nextQuery):
-            set siblingHash to nextQuery.hash
-            remove nextQuery from sortedQueries
-        // 2. sibling is default empty node
-        else if b == 0:
-            set siblingHash to EMPTY_HASH
-        // 3. sibling hash comes from siblingHashes
-        else if b == 1:
-            set siblingHash to the first element of siblingHashes
-            remove the first element of siblingHashes
-
-        let d be the digit at position (h-1) of binaryKey
-        if d == 0:
-            update q.hash to branchHash(q.hash, siblingHash)
-        else if d == 1:
-            update q.hash to branchHash(siblingHash, q.hash)
-
-        remove the first digit of q.binaryBitmap
-        sortedQueries=sortByBitmapAndKey(sortedQueries)
-        sortedQueries=filterQueries(sortedQueries)
-
-    throw error
+    # Finally calculate the root from the provided proof and check that it matches the input Merkle root.
+    return calculate_root(proof.sibling_hashes, filtered_queries) == root
 ```
 
-```java
-function verify(queryKeys, proof, merkleRoot):
-    // check that the proof is valid with respect to queryKeys
-    if length of queryKeys != length of proof.queries:
-        return false
-    for each query q in proof.queries:
-        let k be the corresponding key in queryKeys (the element of queryKeys with same index)
-        // q is an inclusion proof for k or a default empty node
-        if q.key == k:
-            continue
-        // q is an inclusion proof for another leaf node
-        let binaryResponseBitmap = binaryExpansion(q.bitmap) without leading zeros
-        let binaryResponseKey=binaryExpansion(q.key)
-        let binaryQueryKey=binaryExpansion(k)
-        let sharedPrefix = the common prefix of binaryResponseKey and binaryQueryKey
-        if length of binaryResponseBitmap > length of sharedPrefix:
-            // q does not give an non-inclusion proof for k
-            return false
-
-    // calculate the new Merkle root from the proof
-    let calculatedRoot = calculateRoot(proof.siblingHashes, proof.queries)
-    return calculatedRoot == merkleRoot
+```python
+def calculate_root(sibling_hashes: list[bytes], queries: list[Query]) -> bytes:
+    sorted_queries = sorted(queries, key=lambda q: (-q.height, q.key))
+
+    while len(sorted_queries) > 0:
+        q = sorted_queries.pop(0)
+
+        # We reached the top of the tree, return merkle root
+        if q.height == 0:
+            # Check that no extra bits where appended to the bitmap.
+            assert len(q.binary_bitmap) == 0
+            assert len(sibling_hashes) == 0
+            return q.hash
+
+        # we distinguish three cases for the sibling hash:
+        # 1. sibling is next element of sorted_queries
+        if len(sorted_queries) > 0 and are_sibling_queries(q, sorted_queries[0]):
+            sibling_hash = sorted_queries[0].hash
```

Comment:
What is the hash of a query, it is not a property of the `Query` class.

Reply:
fixed